internal/provider/buildEvent.go

```diff
@@ -0,0 +1,240 @@
+package provider
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/kairos-io/kairos-sdk/bus"
+	"github.com/kairos-io/kairos-sdk/types"
+	"github.com/kairos-io/kairos-sdk/utils"
+	"github.com/kairos-io/provider-kairos/v2/internal/services"
+	"github.com/mudler/go-pluggable"
+)
+
+const (
+	K3s = "k3s"
+	K0s = "k0s"
+)
+
+// BuildEvent handles the buildtime event for the provider. Called by kairos-init during the build process.
+func BuildEvent(e *pluggable.Event) pluggable.EventResponse {
+	returnData := pluggable.EventResponse{
+		State: "",
+		Data:  "",
+		Error: "",
+	}
+	l := types.NewKairosLogger("provider-kairos-build", "info", true)
+	l.Logger.Info().Msg("Buildtime event received")
+	l.Logger.Debug().Interface("event", e).Msg("Event details")
+	// unmarshal the event data if needed
+	p := &bus.ProviderPayload{}
+	if e.Data != "" {
+		err := json.Unmarshal([]byte(e.Data), p)
+		if err != nil {
+			l.Logger.Error().Err(err).Msg("Failed to unmarshal event data")
+			returnData.Error = err.Error()
+			returnData.State = bus.EventResponseError
+			return returnData
+		}
+	}
+	// Now move the logger to the requested log level
+	l.SetLevel(p.LogLevel)
+	l.Logger.Debug().Interface("payload", p).Msg("Payload details")
+	// Download the installer script for the provider
+	var url string
+	switch p.Provider {
+	case K3s:
+		url = "https://get.k3s.io"
+	case K0s:
+		url = "https://get.k0s.sh"
+	}
+
+	installerFile := filepath.Join(os.TempDir(), "installer.sh")
+
+	// Download the installer script
+	switch p.Provider {
+	case K3s, K0s:
+		l.Logger.Info().Msgf("Downloading installer script for %s from %s", p.Provider, url)
+		// TODO: Do it with golang instead of needing curl?
+		out, err := exec.Command("curl", "-sfL", url, "-o", installerFile).CombinedOutput()
+		if err != nil {
+			l.Logger.Error().Err(err).Msgf("Failed to download installer script: %s", string(out))
+			returnData.Error = fmt.Sprintf("Failed to download installer script: %s", string(out))
+			returnData.State = bus.EventResponseError
+			return returnData
+		}
+	default:
+		// This is not for us, its for another provider or no provider was specified
+		l.Logger.Info().Msg("No valid provider specified or unsupported provider. Skipping buildtime logic.")
+		returnData.State = bus.EventResponseNotApplicable
+		return returnData
+	}
+	// Make the installer script executable
+	err := os.Chmod(installerFile, 0755)
+	if err != nil {
+		l.Logger.Error().Err(err).Msgf("Failed to make installer script executable: %s", installerFile)
+		returnData.Error = fmt.Sprintf("Failed to make installer script executable: %s", err)
+		returnData.State = bus.EventResponseError
+		return returnData
+	}
+
+	// Install the binaries
+	var out []byte
+	switch p.Provider {
```

Comment:
instead of having the multiple switch cases here, could we have just one and that one calls an instsallation function for the corresponding distro?

Reply:
in order to share things. For example the installer download only downloads the file and sets in a location, generic for both, but the different providers need a different install aprt. so I found ti more easy to split those. Feel free to rework this if you can to account for the diffs easily but I found it to be very confusing in a single swithc as it was repeating a lot of the steps :(

Reply:
hmm i beg to differ, I think having the code like this is what made it so hard to introduce a new k8s distribution in the first place but don't want to block your work, we can have a card to refactor or we try to pair on it if that makes sense ... maybe some mobbing!!!! haha

Reply:
> hmm i beg to differ, I think having the code like this is what made it so hard to introduce a new k8s distribution in the first place but don't want to block your work, we can have a card to refactor or we try to pair on it if that makes sense ... maybe some mobbing!!!! haha

Sounds good. Im not aware of the issues when introducing the k0s operator so maybe ita a good idea to pair on It as you had direct involvement in introducing it and the problems it caused?

Reply:
This is for 3.6 so we have time, although I'm going on holidays after the hack week so maybe you want to take over it if we don't have time to pair. Let's talk next week